SourceKitStressTester/Sources/StressTester/SourceKitDocument.swift

```diff
@@ -381,3 +420,110 @@ public struct SourceState {
     return changed
   }
 }
+
+public struct CompletionMatcher {
+  private let expected: ExpectedResult
+
+  public init(for expected: ExpectedResult) {
+    self.expected = expected
+  }
+
+  /// - returns: true if a match was found
+  public func match(_ result: String, ignoreArgLabels: Bool) -> Bool {
+    if ignoreArgLabels {
+      let name = SwiftName(result)!
+      return name.base == expected.name.base
+    }
+    // Check if the base name and/or argument labels match based on the expected
+    // result kind.
+    return matches(name: result)
+  }
+
+  private func matches(name: String) -> Bool {
+    let resultName = SwiftName(name)!
+    guard resultName.base == expected.name.base else { return false }
+    switch expected.kind {
+    case .call:
+      return name.last == ")" && matchesCall(labels: resultName.argLabels)
+    case .reference:
+      return expected.name.argLabels.isEmpty || expected.name.argLabels == resultName.argLabels
+    case .pattern:
+      // If the expected result didn't match on the associated value: it matches
+      if expected.name.argLabels.isEmpty {
+        return true
+      }
+
+      // Result names for enum cases work differently to functions in that
+      // unlabelled items in the associated values aren't represented, e.g.:
+      //   case foo              // name: foo
+      //   case foo(Int, Int)    // name: foo()
+      //   case foo(x: Int, Int) // name: foo(x:)
+      //   case foo(Int, y: Int) // name: foo(y:)
+
+      // If the result name doesn't have an associated value, but the expected
+      // name does: it doesn't match
+      if name.last != ")" && !expected.name.argLabels.isEmpty {
+        return false
+      }
+
+      // If the expected result bound the entire associated value to a single
+      // unlabelled variable, we're done
+      if expected.name.argLabels == [""] {
+        return true
+      }
+
+      // Otherwise the expected argument labels must either match the
+      // corresponding result arg labels or be "" since they don't have to be
+      // specified when pattern matching.
+      var unmatched = resultName.argLabels[...]
+      return expected.name.argLabels.allSatisfy{ label in
+        if label.isEmpty { return true }
+        if let labelIndex = unmatched.firstIndex(of: label) {
+          unmatched = unmatched.dropFirst(labelIndex - unmatched.startIndex + 1)
+          return true
+        }
+        return false
+      }
+    }
+  }
+
+  private func matchesCall(labels: [String]) -> Bool {
+    var paramLabels = labels[...]
+    var remainingArgLabels = expected.name.argLabels[...]
+
+    guard !paramLabels.isEmpty else { return remainingArgLabels.isEmpty }
+    while let nextParamLabel = paramLabels.popFirst() {
```

Comment:
Instead of slicing the collection and popping elements, you should be able to just iterate over `paramLabels`.